Bug 1129099 - Convert Mozmill js test testAccessLocationBar.js to a Firefox UI Test. r=hskupin

galgeek opened the pull request:
more information: https://bugzilla.mozilla.org/show_bug.cgi?id=1129099
original mozmill test: http://hg.mozilla.org/qa/mozmill-tests/file/default/firefox/tests/functional/testAwesomeBar/testAccessLocationBar.js

conversion begun with PR #70.

whimboo commented on firefox_puppeteer/ui/toolbars.py:
```diff
@@ -331,6 +331,13 @@ def results(self):
+    def selected_index(self):
+        """
+        :returns: The index of the selected autocomplete result.
+        """
```
Please keep the style of the doc as for other properties and add a short description. The :returns: can simply say 'The index'.

whimboo commented on firefox_ui_tests/functional/locationbar/test_access_locationbar.py:
```diff
@@ -0,0 +1,61 @@
+        ]
+        self.test_urls = [self.marionette.absolute_url(t)
+                          for t in self.test_urls]
+        self.test_urls.append('about:blank')
```
Lets move the about:blank page out of this array. We do not have to wait that this has been visited. Make it a separate call to marionette.navigate() in the test.

whimboo commented on firefox_ui_tests/functional/locationbar/test_access_locationbar.py:
```diff
@@ -0,0 +1,61 @@
+        # Clear contents of url bar to focus, then arrow down for list of visited sites
+        # Verify that autocomplete is open and results are displayed
+        self.locationbar.clear()
+        # type key to force autocomplete results to load; Bug 1038614 blur failing OS X 10.10.2
```
So what's with blur() here? I don't see it called anymore before clearing the locationbar, which was our workaround so far to circumvent the bug in Firefox.

whimboo commented on firefox_ui_tests/functional/locationbar/test_access_locationbar.py:
```diff
@@ -0,0 +1,61 @@
+        # Verify that autocomplete is open and results are displayed
+        self.locationbar.clear()
+        # type key to force autocomplete results to load; Bug 1038614 blur failing OS X 10.10.2
+        self.urlbar.send_keys('1')
```
This modifies the pre-conditions for the test. There should be no value in the locationbar when the down key is pressed.

galgeek replied:
The blur() workaround from Bug 1038614 fails for OS X 10.10.2, as I noted in the comment. 
(I stuck this in to be able to test the rest of the test, should've removed...).

whimboo commented on firefox_ui_tests/functional/locationbar/test_access_locationbar.py:
```diff
@@ -0,0 +1,61 @@
+        # Arrow down again to select first item in list, which should appear in this order:
+        #   layout/mozilla_mission.html
+        #   layout/mozilla.html
+        #   layout/mozilla_projects.html
```
Instead of listing them all here, just say something like "reversed order as loaded"

whimboo commented on firefox_ui_tests/functional/locationbar/test_access_locationbar.py:
```diff
@@ -0,0 +1,61 @@
+        #   layout/mozilla_mission.html
+        #   layout/mozilla.html
+        #   layout/mozilla_projects.html
+        # Verify first item is selected and selected item populates location bar with url
```
'First item' should be enough :)

whimboo commented on firefox_ui_tests/functional/locationbar/test_access_locationbar.py:
```diff
@@ -0,0 +1,61 @@
+        # Verify first item is selected and selected item populates location bar with url
+        self.urlbar.send_keys(self.keys.ARROW_DOWN)
+        self.wait_for_condition(lambda _: self.autocomplete_results.selected_index == '0')
+        self.assertIn('mozilla_mission', self.locationbar.value)
```
Please keep what the original test is doing and do not modify conditions. So use 'mission' only, without any 'mozilla' which is also present in all other page titles.

galgeek replied:
Oops, I think this was in first PR, I didn't check it against original. Now corrected.

galgeek commented:
Thanks for your feedback! I believe I've addressed all of it, added a couple of line notes above.
This test fails for me on OS X 10.10.2; the blur() workaround doesn't workaround the bug.
This test gets skipped under xvfb.

whimboo commented on firefox_ui_tests/functional/locationbar/test_access_locationbar.py:
```diff
@@ -0,0 +1,59 @@
+        self.autocomplete_results = self.locationbar.autocomplete_results
+        self.urlbar = self.locationbar.urlbar
+
+    @skip_under_xvfb
```
Do this skip at the class level so that we can avoid issue with tearDown and the autocomplete popup.

whimboo commented on firefox_ui_tests/functional/locationbar/test_access_locationbar.py:
```diff
@@ -0,0 +1,59 @@
+        self.places.wait_for_visited(self.test_urls, load_urls)
+        with self.marionette.using_context('content'):
+            self.marionette.navigate('about:blank')
+        # Need to blur url bar or autocomplete won't load - bug 1038614
```
Please add a blank line above to separate out this code block.

whimboo commented:
This PR also needs a rebase to master.

whimboo commented:
Merged PR as https://github.com/mozilla/firefox-ui-tests/commit/1e1428f9ccb1e74721e55fa8b38d8be8ff99eb1b
